File: libs/langou/path.js
/* ***** BEGIN LICENSE BLOCK *****
 * Distributed under the BSD license:
 *
 * Copyright (c) 2015, xuewen.chu
 * All rights reserved.
 * 
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *     * Redistributions of source code must retain the above copyright
 *       notice, this list of conditions and the following disclaimer.
 *     * Redistributions in binary form must reproduce the above copyright
 *       notice, this list of conditions and the following disclaimer in the
 *       documentation and/or other materials provided with the distribution.
 *     * Neither the name of xuewen.chu nor the
 *       names of its contributors may be used to endorse or promote products
 *       derived from this software without specific prior written permission.
 * 
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL xuewen.chu BE LIABLE FOR ANY
 * DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
 * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
 * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
 * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
 * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 * 
 * ***** END LICENSE BLOCK ***** */

/**************************************************************************/

const _path = requireNative('_path');
const _pkgutil = requireNative('_pkgutil');

/**************************************************************************/

const haveWeb = typeof global.window == 'object';

function split_path(self) {
	if (self._is_split) return;
	self._is_split = true;
	
	var value = self._value;
	var val = '';
	var i = value.indexOf('?');
	
	if (i != -1) {
		val = value.substr(0, i);
		var search = self._search = value.substr(i);
		i = search.indexOf('#');
		if (i != -1) {
			self._search = search.substr(0, i);
			self._hash = search.substr(i);
		}
	} else {
		i = value.indexOf('#');
		if (i != -1) {
			val = value.substr(0, i);
			self._hash = value.substr(i);
		} else {
			val = value;
		}
	}
	self._value = _pkgutil.resolve(val);
}

function parse_base_ext_name(self) {
	if (self._basename == -1) {
		split_path(self);
		var mat = self._value.match(/([^\/\\]+)?(\.[^\/\\\.]+)$|[^\/\\]+$/);
		if (mat) {
			self._basename = mat[0];
			self._extname = mat[2] || '';
		} else {
			self._extname = self._basename = '';
		}
	}
}

function parse_path(self) {
	if (self._is_parse) return;
	self._is_parse = true;
	
	split_path(self);
	
	var mat = self._value.match(/^(([a-z]+:)\/\/)?(([^\/:]+)(?::(\d+))?)?(\/.*)?/);
	if (mat) {
		self._protocol = mat[2] || '';
		self._origin = mat[1] || '';

		if ( mat[3] ) { // http:// or ftp:// or lib://
			self._origin += mat[3];
			self._hostname = mat[4];
			self._port = mat[5] ? mat[5] : '';
		}

		var path = self._filename = mat[6] || '/';
		var i = path.lastIndexOf('/');
		if (i > 0) {
			self._dirname = path.substr(0, i);
		} else {
			self._dirname = '/';
		}
	} else {
		throw new Error(`Parse url error, Illegal URL ${self._value}`);
	}
}

function parse_params(self) {
	if (self._params) 
		return;
	split_path(self);
	
	var params = self._params = { };
	
	if (self._search[0] != '?') 
		return;
		
	var ls = self._search.substr(1).split('&');
	
	for (var i = 0; i < ls.length; i++) {
		var o = ls[i].split('=');
		params[ o[0] ] = o[1] || '';
	}
}

function parse_hash_params(self) {
	if (self._hash_params) 
		return;
	split_path(self);
	
	var params = self._hash_params = { };
	if (self._hash[0] != '#') 
		return;
		
	var ls = self._hash.substr(1).split('&');
	
	for (var i = 0; i < ls.length; i++) {
		var o = ls[i].split('=');
		params[ o[0] ] = o[1] || '';
	}
}

function stringify_params(prefix, params) {
	var rev = [];
	for (var i in params) {
		rev.push(i + '=' + params[i]);
	}
	return rev.length ? prefix + rev.join('&') : '';
}

/**
 * @class URL
 */
class URL {
	
	/**
		* @arg [path] {String}
		* @constructor
		*/
	constructor(path = '') {
		if (!path && haveWeb) {
			path = location.href;
		}
		this._value = path;
	}
	
	// href: "http://xxxx.xxx:81/v1.1.0/./path.js?sasasas&asasasa#sdsdsd"
	get href() {
		parse_path(this);
		return this._origin + this._filename + this._search + this._hash;
	}
	
	/**
		* full path
		* filename: "/D:/Documents/test.js"
		*/
	get filename() {
		parse_path(this);
		return  this._filename;
	}
	
	/**
	 * @get path /a/b/s/test.html?aaaaa=100
	 */
	get path() {
		return this._filename + this._search;
	}
	
	/**
		* full path dir
		* dirname: "/D:/Documents"
		*/
	get dirname() {
		parse_path(this);
		return this._dirname;
	}
	
	// search: "?sasasas&asasasa"
	get search() {
		split_path(this);
		return this._search;
	}
	
	// hash: "#sdsdsd"
	get hash() {
		split_path(this);
		return this._hash;
	}
	
	// host: "langou.org:81"
	get host() {
		parse_path(this);
		return this._hostname + (this._port ? ':' + this._port : '');
	}
	
	// hostname: "langou.org"
	get hostname() {
		parse_path(this);
		return this._hostname;
	}
	
	// origin: "http://langou.org:81"
	get origin() {
		parse_path(this);
		return this._origin;
	}

	// get path base name 
	get basename() {
		parse_base_ext_name(this);
		return this._basename;
	}
	
	// path extname
	get extname() {
		parse_base_ext_name(this);
		return this._extname;
	}
	
	// port: "81"
	get port() {
		parse_path(this);
		return this._port;
	}
	
	// protocol: "http:"
	get protocol() {
		parse_path(this);
		return this._protocol;
	}
	
	get params() {
		parse_params(this);
		return this._params;
	}
	
	get hashParams() {
		parse_hash_params(this);
		return this._hash_params;
	}
	
	// get path param
	getParam(name) {
		return this.params[name];
	}
	
	// set path param
	setParam(name, value) {
		this.params[name] = value || '';
		this._search = stringify_params('?', this._params);
		return this;
	}
	
	// del path param
	deleteParam(name) {
		delete this.params[name];
		this._search = stringify_params('?', this._params);
		return this;
	}
	
	// del all prams
	clearParam() {
		this._params = { };
		this._search = '';
		return this;
	}
	
	// get hash param
	getHash(name) {
		return this.hashParams[name];
	}
	
	// set hash param
	setHash(name, value) {
		this.hashParams[name] = value || '';
		this._hash = stringify_params('#', this._hash_params);
		return this;
	}
	
	// del hash param
	deleteHash(name) {
		delete this.hashParams[name];
		this._hash = stringify_params('#', this._hash_params);
		return this;
	}
	
	// del hash all params
	clearHash(){
		this._hash_params = { };
		this._hash = '';
		return this;
	}
	
	// relative path
	relative(target) {
		target = new URL(target);
		if ( this.origin != target.origin )
			return this._origin + this._filename;
		var ls  = this._filename == '/' ? [] : this._filename.split('/');
		var ls2 = target._filename == '/' ? [] : target._filename.split('/');
		var len = Math.max(ls.length, ls2.length);
		
		for (var i = 1; i < len; i++) {
			if (ls[i] != ls2[i]) {
				len = ls.length - i;
				if (len > 0) {
					for (var j = 0, ls = []; j < len; j++)
						ls.push('..');
					return ls.join('/') + '/' + ls2.splice(i).join('/');
				}
				return ls2.splice(i).join('/');
			}
		}
		return '';
	}
	// @end
}

URL.prototype._is_split = false;
URL.prototype._is_parse = false;
URL.prototype._value = '';
URL.prototype._hostname = '';
URL.prototype._port = '';
URL.prototype._protocol = '';
URL.prototype._search = '';
URL.prototype._hash = '';
URL.prototype._origin = '';
URL.prototype._filename = '';
URL.prototype._dirname = '';
URL.prototype._basename = -1;
URL.prototype._extname = -1;
URL.prototype._params = null;
URL.prototype._hash_params = null;

function get_path(path) {
	return new URL(path);
}

function resolveLocal(...args) {
	return _path.fallbackPath(_pkgutil.resolve(...args));
}

Object.assign(exports, _path);

Object.assign(exports, {

	URL: URL,

	/** 
	 * @func isAbsolute(path) is absolute path
	 */
	isAbsolute: _pkgutil.isAbsolute, // func
	
	/**
	 * @func resolve(path) resolve path 
	 */
	resolve: _pkgutil.resolve, // func

	/**
	 * @func resolveLocal()
	 */
	resolveLocal: resolveLocal,

	/**
	 * full filename
	 */
	basename: function (path) {
		return get_path(path).basename;
	},

	/**
	 * full filename
	 */
	dirname: function (path) {
		return get_path(path).dirname;
	},

	/**
	 * full filename
	 */
	extname: function (path) {
		return get_path(path).extname;
	},

	/**
	 * full filename
	 */
	filename: function (path) {
		return get_path(path).filename;
	},

	/**
	 * full path
	 */
	path: function (path) {
		return get_path(path).path;
	},
	
	search: function (path) {
		return get_path(path).search;
	},
	
	hash: function (path) {
		return get_path(path).hash;
	},
	
	host: function (path) {
		return get_path(path).host;
	},
	
	hostname: function (path) {
		return get_path(path).hostname;
	},
	
	// href origin
	origin: function (path) {
		return get_path(path).origin;
	},
	
	// port: "81"
	port: function (path) {
		return get_path(path).port;
	},
	
	// protocol: "http:"
	protocol: function (path) {
		return get_path(path).protocol;
	},
	
	// href params
	params: function (path) {
		return get_path(path).params;
	},
	
	// hash params 
	hashParams: function (path) {
		return get_path(path).hashParams;
	},
	
	// get path param
	getParam: function (name, path) {
		return get_path(path).getParam(name);
	},
	
	// set path param
	setParam: function (name, value, path) {
		return get_path(path).setParam(name, value).href;
	},
	
	// del path param
	deleteParam: function (name, path) {
		return get_path(path).deleteParam(name).href;
	},
	
	// del all hash param
	clearParam: function (path) {
		return get_path(path).clearParam().href;
	},
	
	// get hash param
	getHash: function (name, path) {
		return get_path(path).getHash(name);
	},
	
	// set hash param
	setHash: function (name, value, path) {
		return get_path(path).setHash(name, value).href;
	},
	
	// del hash param
	deleteHash: function (name, path) {
		return get_path(path).deleteHash(name).href;
	},
	
	// del all hash param
	clearHash: function (path) {
		return get_path(path).clearHash().href;
	},
	
	// relative path
	relative: function (path, target) {
		if (arguments.length > 1) 
			return get_path(path).relative(target);
		else 
			return get_path().relative(path);
	},
	
});

File: flare/render/path.cc
/* ***** BEGIN LICENSE BLOCK *****
 * Distributed under the BSD license:
 *
 * Copyright (c) 2015, xuewen.chu
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *     * Redistributions of source code must retain the above copyright
 *       notice, this list of conditions and the following disclaimer.
 *     * Redistributions in binary form must reproduce the above copyright
 *       notice, this list of conditions and the following disclaimer in the
 *       documentation and/or other materials provided with the distribution.
 *     * Neither the name of xuewen.chu nor the
 *       names of its contributors may be used to endorse or promote products
 *       derived from this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL xuewen.chu BE LIABLE FOR ANY
 * DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
 * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
 * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
 * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
 * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 *
 * ***** END LICENSE BLOCK ***** */

#include "tesselator.h"
#include "./path.h"
#include "../util/handle.h"
#include "../util/bezier.h"

namespace flare {

	PathLine PathLine::Oval(Rect r) {
		float w = r.size.x(), h = r.size.y();
		float x = r.origin.x(), y = r.origin.x();
		float x2 = x + w / 2, y2 = y + h / 2;
		float x3 = x + w, y3 = y + h;
		float cx = w / 2 * 0.552284749831f, cy = h / 2 * 0.552284749831f;

		PathLine path(Vec2(x2, y)); // move
		path.cubic_to(Vec2(x2 + cx, y), Vec2(x3, y2 - cy), Vec2(x3, y2)); // top,right
		path.cubic_to(Vec2(x3, y2 + cy), Vec2(x2 + cx, y3), Vec2(x2, y3)); // right,bottom
		path.cubic_to(Vec2(x2 - cx, y3), Vec2(x, y2 + cy), Vec2(x, y2)); // bottom,left
		path.cubic_to(Vec2(x, y2 - cy), Vec2(x2 - cx, y), Vec2(x2, y)); // left,top
		// path.close_to();

		return path;
	}

	PathLine PathLine::Rect(Rect r) {
		PathLine path(r.origin);
		float x2 = r.origin.x() + r.size.x();
		float y2 = r.origin.y() + r.size.y();
		path.line_to(Vec2(x2, r.origin.y()));
		path.line_to(Vec2(x2, y2));
		path.line_to(Vec2(r.origin.x(), y2));
		path.close_to();
		return path;
	}

	PathLine PathLine::Circle(Vec2 center, float radius) {
		return Oval({ Vec2(center.x() - radius / 2, center.y() - radius / 2), Vec2(radius) * 2 });
	}

	PathLine::PathLine(Vec2 move) {
		add_move(move);
	}

	PathLine::PathLine(Vec2* pts, int len, PathVerb* verbs, int verbsLen) {
		// F_ASSERT(verbs[0] == kVerb_Move);
		_pts.write((float*)pts, -1, len * 2);
		_verbs.write((uint8_t*)verbs, -1, verbsLen);
	}

	PathLine::PathLine() {}

	void PathLine::move_to(Vec2 to) {
		_pts.push(to);
		_verbs.push(kVerb_Move);
	}

	void PathLine::line_to(Vec2 to) {
		// _pts.push(to);
		_pts.write(p1.val, -1, 2);
		_verbs.push(kVerb_Line);
	}

	void PathLine::quad_to(Vec2 control, Vec2 to) {
		_pts.write(control.val, -1, 4);
		_verbs.push(kVerb_Quad);
	}

	void PathLine::cubic_to(Vec2 control1, Vec2 control2, Vec2 to) {
		//_pts.push(control1[0]); _pts.push(control1[1]);
		//_pts.push(control2[0]); _pts.push(control2[1]);
		//_pts.push(to[0]); _pts.push(to[1]);
		_pts.write(control1.val, -1, 6);
		_verbs.push((uint8_t)kVerb_Cubic);
	}

	void PathLine::close_to() {
		_verbs.push((uint8_t)kVerb_Close);
	}

	Array<Vec2> PathLine::to_polygon(int polySize) const {
		//F_ASSERT(_verbs.length());

		TESStesselator* tess = tessNewTess(nullptr);
		ClearScope clear([tess]() { tessDeleteTess(tess); });

		const Vec2* pts = (const Vec2*)*_pts;
		Array<Vec2> polygons, tmpV;
		int len = 0;

		for (auto verb: _verbs) {
			if (len == 0 && verb != kVerb_Move) {
				tmpV.push(Vec2(0)); // use Vec2(0,0) start point
				len++;
			}

			switch(verb) {
				case kVerb_Move:
					if (len > 1) {
						tessAddContour(tess, 2, (float*)&tmpV[tmpV.length() - len], sizeof(Vec2), len);
						len = 1;
					}
					tmpV.push(*pts++);
					break;
				case kVerb_Line:
					tmpV.push(*pts++);
					len++;
					break;
				case kVerb_Quad: { // quadratic
					// F_DEBUG("conic_to:%f,%f|%f,%f", pts[0].x(), pts[0].y(), pts[1].x(), to[1].y());
					QuadraticBezier bezier(tmpV.back(), *pts++, *pts++);
					int sample = PathLine::get_quadratic_bezier_sample(bezier);
					tmpV.extend(tmpV.length() + sample - 1);
					bezier.sample_curve_points(sample, (float*)&tmpV[tmpV.length() - sample]);
					len += sample - 1;
					break;
				}
				case kVerb_Cubic: {// cubic
					//  F_DEBUG("cubic_to:%f,%f|%f,%f|%f,%f",
					//           pts[0].x(), pts[0].y(), pts[1].x(), to[1].y(), pts[2].x(), to[2].y());
					CubicBezier bezier(tmpV.back(), *pts++, *pts++, *pts++);
					int sample = PathLine::get_cubic_bezier_sample(bezier);
					tmpV.extend(tmpV.length() + sample - 1);
					bezier.sample_curve_points(sample, (float*)&tmpV[tmpV.length() - sample]);
					len += sample - 1;
					break;
				}
				default: // close
					if (len) {
						tmpV.push(tmpV[tmpV.length() - len++]);
						tessAddContour(tess, 2, (float*)&tmpV[tmpV.length() - len], sizeof(Vec2), len);
						len = 0;
					}
					break;
			}
		}

		if (len > 1) { // closure
			tessAddContour(tess, 2, (float*)&tmpV[tmpV.length() - len], sizeof(Vec2), len);
		}

		// Convert to convex contour vertex data
		if ( tessTesselate(tess, TESS_WINDING_POSITIVE,
											TESS_CONNECTED_POLYGONS/*TESS_POLYGONS*/, polySize, 2, 0) 
		) {
			const int nelems = tessGetElementCount(tess);
			const TESSindex* elems = tessGetElements(tess);
			const TESSreal* verts = tessGetVertices(tess);
			for (int i = 0; i < nelems * polySize; i++) {
				float x = verts[elems[0]];
				float y = verts[elems[1]];
				polygons.push(Vec2(x, y));
				elems += 2;
			}
		}

		return polygons;
	}

	Array<Vec2> PathLine::to_edge_line() const {
		//F_ASSERT(_verbs.length());

		const Vec2* pts = ((const Vec2*)*_pts) - 1;
		Array<Vec2> edges;
		int len = 0;
		bool isZeor = true;

		for (auto verb: _verbs) {

			switch(verb) {
				case kVerb_Move:
					pts++;
					len = 0;
					isZeor = false;
					break;
				case kVerb_Line:
					edges.push(isZeor ? (pts++, Vec2()): *pts++); edges.push(*pts); // edge 0
					len+=2;
					isZeor = false;
					break;
				case kVerb_Quad: { // Quadratic
					//  F_DEBUG("conic_to:%f,%f|%f,%f", pts[0].x(), pts[0].y(), pts[1].x(), to[1].y());
					QuadraticBezier bezier(isZeor ? (pts++, Vec2()): *pts++, *pts++, *pts++);
					int sample = PathLine::get_quadratic_bezier_sample(bezier);
					auto points = bezier.sample_curve_points(sample);
					for (int i = 0; i < sample - 1; i++) {
						edges.push(points[i]); edges.push(points[i + 1]); // add edge line
						len+=2;
					}
					isZeor = false;
					break;
				}
				case kVerb_Cubic: { // cubic
					//  F_DEBUG("cubic_to:%f,%f|%f,%f|%f,%f",
					//           pts[0].x(), pts[0].y(), pts[1].x(), to[1].y(), pts[2].x(), to[2].y());
					CubicBezier bezier(isZeor ? (pts++, Vec2()): *pts++, *pts++, *pts++, *pts++);
					int sample = PathLine::get_cubic_bezier_sample(bezier);
					auto points = bezier.sample_curve_points(sample);
					for (int i = 0; i < sample - 1; i++) {
						edges.push(points[i]); edges.push(points[i + 1]); // add edge line
						len+=2;
					}
					isZeor = false;
					break;
				}
				default: // close
					if (len) {
						edges.push(*pts); edges.push(edges[edges.length() - len - 1]); // add close edge line
					}
					len = 0;
					isZeor = true;
					break;
			}
		}

		return edges;
	}

	void PathLine::transfrom(const Mat& matrix) {
		float* pts = *_pts;
		float* e = pts + _pts.length();
		while (pts < e) {
			*((Vec2*)pts) = matrix * (*(Vec2*)pts);
			pts += 2;
		}
	}

	void PathLine::scale(Vec2 scale) {
		float* pts = *_pts;
		float* e = pts + _pts.length();
		while (pts < e) {
			pts[0] *= scale[0];
			pts[1] *= scale[1];
			pts += 2;
		}
	}

	PathLine PathLine::normalized() const {
		//F_ASSERT(_verbs.length());

		const Vec2* pts = ((const Vec2*)*_pts);
		PathLine line;
		bool isZeor = true;

		for (auto verb: _verbs) {
			switch(verb) {
				case kVerb_Move:
					line.move_to(*pts++);
					isZeor = false;
					break;
				case kVerb_Line:
					if (isZeor)
						line.move_to(Vec2()); // add zeor
					line.add_line(*pts++);
					isZeor = false;
					break;
				case kVerb_Quad: { // quadratic
					if (isZeor)
						line.move_to(Vec2());
					QuadraticBezier bezier(line.back(), *pts++, *pts++);
					int sample = PathLine::get_quadratic_bezier_sample(bezier) - 1;
					line._pts.extend(line._pts.length() + sample * 2);
					bezier.sample_curve_points(sample, &line._pts[line._pts.length() - sample * 2 - 2]);
					line._verbs.extend(line._verbs.length() + sample);
					memset(line._verbs.val() + (line._verbs.length() - sample), kVerb_Line, sample);
					isZeor = false;
					break;
				}
				case kVerb_Cubic: { // cubic
					if (isZeor)
						line.move_to(Vec2());
					CubicBezier bezier(line.back(), *pts++, *pts++, *pts++);
					int sample = PathLine::get_cubic_bezier_sample(bezier) - 1;
					line._pts.extend(line._pts.length() + sample * 2);
					bezier.sample_curve_points(sample, &line._pts[line._pts.length() - sample * 2 - 2]);
					line._verbs.extend(line._verbs.length() + sample);
					memset(line._verbs.val() + (line._verbs.length() - sample), kVerb_Line, sample);
					isZeor = false;
					break;
				}
				default: // close
					line._verbs.push(kVerb_Close);
					isZeor = true;
					break;
			}
		}

		return line;
	}

	PathLine PathLine::clip(const PathLine& path) const {
		// TODO ...
		return PathLine();
	}

	int PathLine::get_quadratic_bezier_sample(const QuadraticBezier& curve) const {
		// TODO ...
		return 10;
	}

	int PathLine::get_cubic_bezier_sample(const CubicBezier& curve) const {
		// TODO ...
		return 10;
	}

}